sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/ReadFromTasks.java

```diff
@@ -0,0 +1,84 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.iceberg;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import org.apache.beam.sdk.io.range.OffsetRange;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.Row;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.io.CloseableIterable;
+
+/**
+ * Bounded read implementation.
+ *
+ * <p>For each {@link ReadTask}, reads Iceberg {@link Record}s, and converts to Beam {@link Row}s.
+ *
+ * <p>Implemented as an SDF to leverage communicating bundle size (i.e. {@link DoFn.GetSize}) to the
+ * runner, to help with scaling decisions.
+ */
+@DoFn.BoundedPerElement
+class ReadFromTasks extends DoFn<KV<ReadTaskDescriptor, ReadTask>, Row> {
+  private final IcebergScanConfig scanConfig;
+  private final Counter scanTasksCompleted =
+      Metrics.counter(ReadFromTasks.class, "scanTasksCompleted");
+
+  ReadFromTasks(IcebergScanConfig scanConfig) {
+    this.scanConfig = scanConfig;
+  }
+
+  @ProcessElement
+  public void process(
+      @Element KV<ReadTaskDescriptor, ReadTask> element,
+      RestrictionTracker<OffsetRange, Long> tracker,
+      OutputReceiver<Row> out)
+      throws IOException, ExecutionException, InterruptedException {
+    if (!tracker.tryClaim(0L)) {
+      return;
+    }
+    ReadTask readTask = element.getValue();
+    Table table =
+        TableCache.get(scanConfig.getTableIdentifier(), scanConfig.getCatalogConfig().catalog());
+
+    FileScanTask task = readTask.getFileScanTask();
+    try (CloseableIterable<Record> reader = ReadUtils.createReader(task, table)) {
+      for (Record record : reader) {
+        Row row = IcebergUtils.icebergRecordToBeamRow(scanConfig.getSchema(), record);
+        out.output(row);
+      }
+    }
+    scanTasksCompleted.inc();
+  }
+
+  @GetSize
+  public double getSize(@Element KV<ReadTaskDescriptor, ReadTask> element) {
+    return element.getValue().getByteSize();
```

Comment:
this may work better for autoscaling if it is consistent with the output bytes.  I'm guessing this byte size is more file-bytes and not necessarily the record bytes after conversion to beam rows. See sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/estimator/BytesThroughputEstimator.java for what was needed for Spanner change streams.

Could add a TODO

Reply:
Yep this is currently just the compressed Parquet file size. 

I see KafkaIO keeps a running tracker for average record size: [ReadFromKafkaDoFn.java](https://github.com/apache/beam/blob/d6a5e730ad95973c762ef8f6ea4958592aa1cb6b/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java#L512-L516)

We could take inspo from that, but implementing it here might get tricky because AFAIK there's no straightforward way to determine the size of a Beam Row (unlike Kafka records which are raw bytes).

Reply:
Added a TODO